File: ruby/src/final-prices-with-a-special-discount-in-a-shop.rb
# @param {Integer[]} prices
# @return {Integer[]}
def final_prices(prices)
  stack = []
  final = Array.new(prices.size)
  (prices.length - 1).downto(0) do |i|
    while stack.any? && stack.last > prices[i]
      stack.pop
    end
    final[i] = prices[i] - (stack.last || 0)
    stack.push(prices[i])
  end
  final
end

final_prices([8,4,6,2,3]) => [4,2,4,2,3]
final_prices([1,2,3,4,5]) => [1,2,3,4,5]
final_prices([10,1,1,6]) => [9,0,1,6]

File: elixir/final-prices-with-a-special-discount-in-a-shop.ex
defmodule Solution do
  @spec final_prices(prices :: [integer]) :: [integer]
  def final_prices(prices), do: recur(prices) |> elem(0)

  def recur([]), do: {[], [0]}
  def recur([p | prices]) do
    {result, stack} = recur(prices)
    stack = Enum.drop_while(stack, fn item -> item > p end)
    {[p - hd(stack) | result], [p | stack]}
  end
end
